### ISSUE TITLE
Token expiration 

### ISSUE THREAD

**Issue by greberg** [radical-squared/aquatemp#53]
**Describe the bug**
Not sure how to manage this but it looks like the token expires if you for instance login in the app. This mean you need to reload the integration to get a new token and with this it also log out the app as well. Not sure if its possible to get a life long token or at least not a new token for every login

**To Reproduce**

1. Go to 'your IOS/android app and login. You will start to get Result: {'error_msg': '请重新登录', 'error_msg_code': '', 'isReusltSuc': False, 'objectResult': None, 'error_code': '-100'} in log in HomeAssistant
2. Reload integration and you will directly be logged out in app and integration starts working again

**Comment by davidjac** [comment]
Yes I am experiencing same. Seems to not update until reload - also noticed App logged out.

**Comment by elad-bar** [comment]
It should relogin once needed, do you have logs with debug level to share?
First attempt should fail, it will relogin on next attempt to update

Thanks

**Comment by elad-bar** [comment]
The behaviour of using one device is by the server, nothing we can do in the integration, just will relogin in case you have logged in to the app and disconnected the integration as rsult from it

**Comment by greberg** [comment]
Yes, my thoughts as well. Maybe worth adding to the readme. 

Would it be possible in some way notify in HA if the token has expired so you can trigger a reload of the integration in an automation? Because we don't want to stress the system, for instance if you logging with aquatemp in your phone you don't want to relogin direct in HA and trigger an logout in the app and so on. But at the same time I don't want to go into HA and reload the integration every time I have used the app so my automations work with the pump. Any ideas here? (You can always argue if I should use the aquatemp app anymore :-))

**Comment by davidjac** [comment]
If its a matter of only one device logged in I will logout of the app (please confirm) I suspect I can do all I need, monitor temp, set timer  etc via your integration.

**Comment by davidjac** [comment]
Actually just noticed app comes up with a new warning you are logged in elsewhere.  So at least there's clarity.

**Comment by elad-bar** [comment]
yes, in the app you are getting a popup message,
I will add it to README as well, will leave the issue open up until I will add that clearification

thanks for raising it

**Comment by greberg** [comment]
Problem is when you lose the token in the integration it is no way to actually know if it has expire. Only way to know it either to reload it just because of or check in the app. So question would be if its possible to add a sensor specific for the integration which pings the api and give available or non available based on the error message from the api. This line in the logs gives me the answer today that my token have dropped

2023-06-28 08:29:20.507 ERROR (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Error fetching parameters, Error: 'NoneType' object is not iterable, Line: 165

**Comment by elad-bar** [comment]
Ping the api is exactly what is being done, but the problem as you describe it is that the first attempt will fail for sure,
And it seems that if you are getting that error something is not protected enough,
What I'm think might solve it is when it fails on expiration, try to recover automatically without throwing the erro, if it will fail on this attempt it will log it, wdyt?

**Comment by greberg** [comment]
To possible scenario I can see is that it will start conflicting with the app. So direct if you have login in the app you will be logged out in the component and then the component will login and then you will be directly logged out in the app. So that was a little why I proposed to create a valid token sensor. Those that want to to solve this can then create a automation that re-run the component and creates a new token and those that maybe will create a automation that actually delay the re-run of the component so the can use the app for a while and then let the automation take care of the re-run. What do you think about that solution?

**Comment by elad-bar** [comment]
Basically what you are saying is that if the component was not update over the last 30 seconds, it means you are offline, you have thay data already now,
Having a sensor like that is usually being implemented when you have live connection like websocket / mqtt, not with http web service, for that there is unavailable state,
I can add a configuration parameter, if it is enabled it will auto relogin, otherwise, will leave it to create custom automation

**Comment by greberg** [comment]
Hmm, that is the thing. If this occur, the component still says online. The re-initiating of the token doesn't work if this is happen. At least not for me :-). So if Im logged in in the component and then login in the app the connection is broken but the component doesn't pick that up.

**Comment by elad-bar** [comment]
please check the new release [v3.0.14](https://github.com/radical-squared/aquatemp/releases/tag/v3.0.14),
Added new sensor of `API Status`, will get disconnected state when the integration got "disconnection" error (401), it will also try to reauthenticate, so the expectation that you will just see that it got disconnected and connected again, but if it's not working, you can base automation upon state change,
pls let me know how it works for you

thanks

**Comment by greberg** [comment]
Cool! Will try it out for a couple of days to see how it works out. Thanks!

**Comment by greberg** [comment]
Sorry, doesn't seems it worked. Just received 188 occurrences and counting of this:
Logger: custom_components.aqua_temp.managers.aqua_temp_api
Source: custom_components/aqua_temp/managers/aqua_temp_api.py:166
Integration: Aqua Temp ([documentation](https://github.com/radical-squared/aquatemp), [issues](https://github.com/radical-squared/aquatemp/issues))
First occurred: 08:53:02 (188 occurrences)
Last logged: 10:26:32

Failed to update (Attempt #3), Error: 'NoneType' object is not iterable, Line: 177

Looks like the update service doesn't work:
2023-07-04 10:22:32.210 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Request to https://cloud.linked-go.com:449/crmservice/api/app/device/getDataByCode, Body: {'deviceCode': 'hidden', 'protocalCodes': ['Mode', 'Power', 'Mode', 'Manual-mute', 'Set_Temp', '2074', '2075', '2076', '2077', '1158', '1159', '2064', 'ver', 'software_code', 'D01', 'D01-bar', 'D02', 'D03', 'D04', 'D06', 'D07', 'D08', 'D08-bar', 'D09', 'D10', 'D10-bar', 'E01', 'E02', 'E03', 'E04', 'E05', 'E06', 'E07', 'E09', 'E10', 'E11', 'E12', 'F01', 'F02', 'F02-bar', 'F03', 'F03-bar', 'F04', 'F04-bar', 'F05', 'F05-bar', 'F06', 'F06-bar', 'F07', 'F07-bar', 'F10', 'F11', 'F11-r', 'F12', 'F12-r', 'F13', 'F13-r', 'F14', 'F15', 'F16', 'F16-r', 'F17', 'F18', 'F19', 'F19-r', 'F20', 'H01', 'H02', 'H03', 'H06', 'H07', 'H08', 'H09', 'H10', 'H11', 'H12', 'H13', 'H14', 'H15', 'H16', 'H17', 'H18', 'H19', 'H20', 'H21', 'H22', 'H23', 'H24', 'H25', 'H26', 'H27', 'H28', 'H29', 'H30', 'H31', 'H32', 'H33', 'H34', 'H35', 'H36', 'H37', 'H38', 'H39', 'H40', 'H41', 'H42', 'H43', 'O01', 'O02', 'O03', 'O04', 'O05', 'O06', 'O07', 'O08', 'O09', 'P01', 'P02', 'P03', 'P04', 'P05', 'P06', 'P07', 'P08', 'P09', 'R01', 'R02', 'R03', 'R04', 'R05', 'R08', 'R09', 'R10', 'R11', 'R12', 'S01', 'S02', 'S03', 'S04', 'S05', 'S06', 'T01', 'T02', 'T03', 'T04', 'T05', 'T06', 'T07', 'T08', 'T09', 'T10', 'T11', 'T12', 'T13', 'T14', 'T15', 'T16', 'T17', 'T18', 'T19', 'T20', 'T21', 'T22', 'T23', 'T24', 'T25', 'T26', 'T27', 'T28']}, Result: {'error_msg': '请重新登录', 'error_msg_code': '', 'isReusltSuc': False, 'objectResult': None, 'error_code': '-100'}

The API sensor says connected unfortunately

**Comment by elad-bar** [comment]
this is not related to the issue (reported as #55 ), 
since I didn't get that error, can you please add log with debug level (in #55)?

thanks

**Comment by greberg** [comment]
Hi, no I know this is related to the added API sensor that you did for me in this issue :-). that shows if the token is expired or not. Only way of getting the integration up and running again was to reload the integration.

**Comment by elad-bar** [comment]
@greberg what is the status for that issue after version [v3.0.17](https://github.com/radical-squared/aquatemp/releases/tag/v3.0.17)?

thanks

**Comment by greberg** [comment]
Need to come to keep it open a little longer because I have updated some other stuff in HA so I have restarted HA and with that also restarted the integration. I will keep you updated

**Comment by greberg** [comment]
So the token expired yesterday and didn't recovered. Unfortunately the api status sensor didn't recognised that. 
<img width="562" alt="Screenshot 2023-07-09 at 10 14 58" src="https://github.com/radical-squared/aquatemp/assets/213160/971d10d3-a939-43f4-97d1-9baa74a4b55b">


Log from HA doesn't say that much:
  
2023-07-08 16:14:23.918 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_coordinator] Finished fetching Aqua Temp data in 1.479 seconds (success: True)
2023-07-08 16:14:23.919 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Target temp PC R02, HA Mode: heat
2023-07-08 16:14:23.919 DEBUG (MainThread) [custom_components.aqua_temp.climate] _attr_hvac_mode: heat
2023-07-08 16:14:23.919 DEBUG (MainThread) [custom_components.aqua_temp.climate] _attr_target_temperature: 30.0
2023-07-08 16:14:23.919 DEBUG (MainThread) [custom_components.aqua_temp.climate] _attr_fan_mode: auto
2023-07-08 16:14:23.919 DEBUG (MainThread) [custom_components.aqua_temp.climate] _attr_min_temp: 8.0
2023-07-08 16:14:23.919 DEBUG (MainThread) [custom_components.aqua_temp.climate] _attr_max_temp: 28.0
2023-07-08 16:14:53.440 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Starting to update device: 34EAE79F5C1E
2023-07-08 16:14:53.592 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Request to https://cloud.linked-go.com:449/crmservice/api/app/device/getDataByCode, Body: {'deviceCode': '34EAE79F5C1E', 'protocalCodes': ['Mode', 'Power', 'Mode', 'Manual-mute', 'Set_Temp', '2074', '2075', '2076', '2077', '1158', '1159', '2064', 'ver', 'software_code', 'D01', 'D01-bar', 'D02', 'D03', 'D04', 'D06', 'D07', 'D08', 'D08-bar', 'D09', 'D10', 'D10-bar', 'E01', 'E02', 'E03', 'E04', 'E05', 'E06', 'E07', 'E09', 'E10', 'E11', 'E12', 'F01', 'F02', 'F02-bar', 'F03', 'F03-bar', 'F04', 'F04-bar', 'F05', 'F05-bar', 'F06', 'F06-bar', 'F07', 'F07-bar', 'F10', 'F11', 'F11-r', 'F12', 'F12-r', 'F13', 'F13-r', 'F14', 'F15', 'F16', 'F16-r', 'F17', 'F18', 'F19', 'F19-r', 'F20', 'H01', 'H02', 'H03', 'H06', 'H07', 'H08', 'H09', 'H10', 'H11', 'H12', 'H13', 'H14', 'H15', 'H16', 'H17', 'H18', 'H19', 'H20', 'H21', 'H22', 'H23', 'H24', 'H25', 'H26', 'H27', 'H28', 'H29', 'H30', 'H31', 'H32', 'H33', 'H34', 'H35', 'H36', 'H37', 'H38', 'H39', 'H40', 'H41', 'H42', 'H43', 'O01', 'O02', 'O03', 'O04', 'O05', 'O06', 'O07', 'O08', 'O09', 'P01', 'P02', 'P03', 'P04', 'P05', 'P06', 'P07', 'P08', 'P09', 'R01', 'R02', 'R03', 'R04', 'R05', 'R08', 'R09', 'R10', 'R11', 'R12', 'S01', 'S02', 'S03', 'S04', 'S05', 'S06', 'T01', 'T02', 'T03', 'T04', 'T05', 'T06', 'T07', 'T08', 'T09', 'T10', 'T11', 'T12', 'T13', 'T14', 'T15', 'T16', 'T17', 'T18', 'T19', 'T20', 'T21', 'T22', 'T23', 'T24', 'T25', 'T26', 'T27', 'T28']}, Result: {'error_msg': '请重新登录', 'error_msg_code': '', 'isReusltSuc': False, 'objectResult': None, 'error_code': '-100'}
2023-07-08 16:14:53.592 ERROR (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Failed to update device 34EAE79F5C1E, Error: Invalid token, Flow: Fetch data for device 34EAE79F5C1E, Line: 227
2023-07-08 16:14:53.600 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_coordinator] Finished fetching Aqua Temp data in 0.160 seconds (success: True)

**Comment by dodg3r** [comment]
Hi.

There is a function in the app so you can share your heatpump to your friends/family.
Then you can create two accounts and use one of them for homeassistant.

**Comment by elad-bar** [comment]
@greberg it seems that the token expired and relogin took place in less than 8 nanoseconds, means that less than 3 retries applied, no need to change the state of the sensor in that case,
That means it actually worked according to design

**Comment by elad-bar** [comment]
@dodg3r not sure I understand, can you pls explain what are you trying to solve?

Thanks

**Comment by kopierschnitte** [comment]
I've already done this before: Just create a 2nd account and share the device with this account. This way, you can use the app with those credentials and the HA connection (main account) won't get kicked.

**Comment by elad-bar** [comment]
Ah... nice!

**Comment by greberg** [comment]
> @greberg it seems that the token expired and relogin took place in less than 8 nanoseconds, means that less than 3 retries applied, no need to change the state of the sensor in that case, That means it actually worked according to design

The problem is that it didn't succeed with the relogin so it continue like this for 1400 times in the log. So what I want is that after the 3 times of retries set the API status to disconnect.

**Comment by elad-bar** [comment]
I will need logs in debug level,
Are you using the app in parallel? Because 1400 is very close to 1440 which is the number of updates that integration produces a day,
It could indicate that something is fighting on generating different token that invalidates the token of the integration,
If that's the case, I can't do anything with it,
I set 3 retries before it change the sensor to disconnected, in each of the times it tries to revive the token

**Comment by greberg** [comment]
No not using the app so it can't be that unfortunately. Very strange with the three re-tries because I can't see it actually succeed at all after it lost the token. I have this in the configuration file, doesn't that mean I will get debug for the app?
logger:
  default: error
  logs:
    custom_components.aqua_temp: debug

The only logs that I have for you is the once I posted to you earlier

**Comment by elad-bar** [comment]
Checked the code again, when it fails 3 time the message include the attempt count, so you should have 3 times tge error you got then additional with the attempt count.

Can you check if the app is working for you? 
3 days ago my device disappeared from the app, i needed to register it again and since then it works perfectly, in addition, i get the model name,
Could be another update from Aqua Temp

**Comment by elad-bar** [comment]
@dodg3r @kopierschnitte thanks for the suggestion, working great!
in addition, found out that using the shared device, API provides more details about the device, including the model name as listed in the product not model of manufacturer.

**Comment by davidjac** [comment]
Can you clarify the sharing process. I put an alternative email address into add member > device share. But recieve no email. Am I in the correct place?

**Comment by elad-bar** [comment]
Get to the other account, you will get in app message that will ask you to accept the device invitation

**Comment by greberg** [comment]
> Checked the code again, when it fails 3 time the message include the attempt count, so you should have 3 times tge error you got then additional with the attempt count.
> 
> Can you check if the app is working for you? 3 days ago my device disappeared from the app, i needed to register it again and since then it works perfectly, in addition, i get the model name, Could be another update from Aqua Temp

The app is working. I have a easy way of reproducing the error now as well. So this is what happens:
1. I have the integration up and running. Works great
2. After a while the API token seems to expire (You can recreate it by log in using the app then the token will expire directly in the integration. Please remember to log out in the app so the integration can create a new token itself)
3. Here I would expect the integration create a new token and then start working but I get this in the log repeatedly: Failed to update device 34EAE79F5C1E, Error: Cannot connect to host cloud.linked-go.com:449 ssl:False [Try again], Line: 227
Failed to update device 34EAE79F5C1E, Error: Invalid token, Flow: Fetch data for device 34EAE79F5C1E, Line: 227
4. Also the the sensor: API status still shows connected which should show disconnected

Only way of getting the integration up and running is by reloading the integration or restarting HA

Log is attached:
 
[home-assistant_aqua_temp_2023-07-10T07-50-47.510Z.log](https://github.com/radical-squared/aquatemp/files/11999270/home-assistant_aqua_temp_2023-07-10T07-50-47.510Z.log)

**Comment by davidjac** [comment]
> Get to the other account, you will get in app message that will ask you to accept the device invitation

You mean I should recieve an email to the shared account? Or do I set up an in app account using that new email address. If the former I get no message.

**Comment by elad-bar** [comment]
> > Checked the code again, when it fails 3 time the message include the attempt count, so you should have 3 times tge error you got then additional with the attempt count.
> > Can you check if the app is working for you? 3 days ago my device disappeared from the app, i needed to register it again and since then it works perfectly, in addition, i get the model name, Could be another update from Aqua Temp
> 
> The app is working. I have a easy way of reproducing the error now as well. So this is what happens:
> 
> 1. I have the integration up and running. Works great
> 2. After a while the API token seems to expire (You can recreate it by log in using the app then the token will expire directly in the integration. Please remember to log out in the app so the integration can create a new token itself)
> 3. Here I would expect the integration create a new token and then start working but I get this in the log repeatedly: Failed to update device 34EAE79F5C1E, Error: Cannot connect to host cloud.linked-go.com:449 ssl:False [Try again], Line: 227
>    Failed to update device 34EAE79F5C1E, Error: Invalid token, Flow: Fetch data for device 34EAE79F5C1E, Line: 227
> 4. Also the the sensor: API status still shows connected which should show disconnected
> 
> Only way of getting the integration up and running is by reloading the integration or restarting HA
> 
> Log is attached:
> 
> [home-assistant_aqua_temp_2023-07-10T07-50-47.510Z.log](https://github.com/radical-squared/aquatemp/files/11999270/home-assistant_aqua_temp_2023-07-10T07-50-47.510Z.log)

That's the expected behavior as I described, error is being presented but it recovers as part of the 3 attempts, the problem you are describing is that it happens 1400 times a day, which means that something is logging in to the app and terminates the token of the integration which forces it to relogin (the error message you see),
Nothing I can do with it but changing the log level of the message to debug and log the error only after the 3rd failed attempt

**Comment by greberg** [comment]
The strange thing though is that if I reload the integration it starts to work directly again so something else is happening compared to reload of the integration and the reinitiating of the token.

**Comment by elad-bar** [comment]
> > Get to the other account, you will get in app message that will ask you to accept the device invitation
> 
> You mean I should recieve an email to the shared account? Or do I set up an in app account using that new email address. If the former I get no message.

The entire process:
- Create account 1
- Login
- Logout
- Create account 2
- Login
- Logout
- Login to account 1
- Add device
- Share device to the email of account 2
- Logout
- Login to account 2
- Go to messages
- Confirm invitation from account 1

**Comment by jonml** [comment]
> The strange thing though is that if I reload the integration it starts to work directly again so something else is happening compared to reload of the integration and the reinitiating of the token.

hi,
this is happening to me too.  im running the latest HA, and 3.0.17 of this integration.

i have set up the aqua temp app to have a login that is only used for Home Assistant.  This works fine for between 1 and 2 days, then it fails to connect.

Here is an extract from the debug logs:

2023-07-12 09:40:31.251 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Starting to update device: 289C6E20F72C
2023-07-12 09:40:31.349 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Request to https://cloud.linked-go.com:449/crmservice/api/app/device/getDataByCode, Body: {'deviceCode': '289C6E20F72C', 'protocalCodes': ['Mode', 'Power', 'Mode', 'Manual-mute', 'Set_Temp', '2074', '2075', '2076', '2077', '1158', '1159', '2064', 'ver', 'software_code', 'D01', 'D01-bar', 'D02', 'D03', 'D04', 'D06', 'D07', 'D08', 'D08-bar', 'D09', 'D10', 'D10-bar', 'E01', 'E02', 'E03', 'E04', 'E05', 'E06', 'E07', 'E09', 'E10', 'E11', 'E12', 'F01', 'F02', 'F02-bar', 'F03', 'F03-bar', 'F04', 'F04-bar', 'F05', 'F05-bar', 'F06', 'F06-bar', 'F07', 'F07-bar', 'F10', 'F11', 'F11-r', 'F12', 'F12-r', 'F13', 'F13-r', 'F14', 'F15', 'F16', 'F16-r', 'F17', 'F18', 'F19', 'F19-r', 'F20', 'H01', 'H02', 'H03', 'H06', 'H07', 'H08', 'H09', 'H10', 'H11', 'H12', 'H13', 'H14', 'H15', 'H16', 'H17', 'H18', 'H19', 'H20', 'H21', 'H22', 'H23', 'H24', 'H25', 'H26', 'H27', 'H28', 'H29', 'H30', 'H31', 'H32', 'H33', 'H34', 'H35', 'H36', 'H37', 'H38', 'H39', 'H40', 'H41', 'H42', 'H43', 'O01', 'O02', 'O03', 'O04', 'O05', 'O06', 'O07', 'O08', 'O09', 'P01', 'P02', 'P03', 'P04', 'P05', 'P06', 'P07', 'P08', 'P09', 'R01', 'R02', 'R03', 'R04', 'R05', 'R08', 'R09', 'R10', 'R11', 'R12', 'S01', 'S02', 'S03', 'S04', 'S05', 'S06', 'T01', 'T02', 'T03', 'T04', 'T05', 'T06', 'T07', 'T08', 'T09', 'T10', 'T11', 'T12', 'T13', 'T14', 'T15', 'T16', 'T17', 'T18', 'T19', 'T20', 'T21', 'T22', 'T23', 'T24', 'T25', 'T26', 'T27', 'T28']}, Result: {'error_msg': '请重新登录', 'error_msg_code': '', 'isReusltSuc': False, 'objectResult': None, 'error_code': '-100'}
2023-07-12 09:40:31.349 ERROR (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Failed to update device 289C6E20F72C, Error: Invalid token, Flow: Fetch data for device 289C6E20F72C, Line: 227
2023-07-12 09:41:01.251 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Starting to update device: 289C6E20F72C
2023-07-12 09:41:01.348 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Request to https://cloud.linked-go.com:449/crmservice/api/app/device/getDataByCode, Body: {'deviceCode': '289C6E20F72C', 'protocalCodes': ['Mode', 'Power', 'Mode', 'Manual-mute', 'Set_Temp', '2074', '2075', '2076', '2077', '1158', '1159', '2064', 'ver', 'software_code', 'D01', 'D01-bar', 'D02', 'D03', 'D04', 'D06', 'D07', 'D08', 'D08-bar', 'D09', 'D10', 'D10-bar', 'E01', 'E02', 'E03', 'E04', 'E05', 'E06', 'E07', 'E09', 'E10', 'E11', 'E12', 'F01', 'F02', 'F02-bar', 'F03', 'F03-bar', 'F04', 'F04-bar', 'F05', 'F05-bar', 'F06', 'F06-bar', 'F07', 'F07-bar', 'F10', 'F11', 'F11-r', 'F12', 'F12-r', 'F13', 'F13-r', 'F14', 'F15', 'F16', 'F16-r', 'F17', 'F18', 'F19', 'F19-r', 'F20', 'H01', 'H02', 'H03', 'H06', 'H07', 'H08', 'H09', 'H10', 'H11', 'H12', 'H13', 'H14', 'H15', 'H16', 'H17', 'H18', 'H19', 'H20', 'H21', 'H22', 'H23', 'H24', 'H25', 'H26', 'H27', 'H28', 'H29', 'H30', 'H31', 'H32', 'H33', 'H34', 'H35', 'H36', 'H37', 'H38', 'H39', 'H40', 'H41', 'H42', 'H43', 'O01', 'O02', 'O03', 'O04', 'O05', 'O06', 'O07', 'O08', 'O09', 'P01', 'P02', 'P03', 'P04', 'P05', 'P06', 'P07', 'P08', 'P09', 'R01', 'R02', 'R03', 'R04', 'R05', 'R08', 'R09', 'R10', 'R11', 'R12', 'S01', 'S02', 'S03', 'S04', 'S05', 'S06', 'T01', 'T02', 'T03', 'T04', 'T05', 'T06', 'T07', 'T08', 'T09', 'T10', 'T11', 'T12', 'T13', 'T14', 'T15', 'T16', 'T17', 'T18', 'T19', 'T20', 'T21', 'T22', 'T23', 'T24', 'T25', 'T26', 'T27', 'T28']}, Result: {'error_msg': '请重新登录', 'error_msg_code': '', 'isReusltSuc': False, 'objectResult': None, 'error_code': '-100'}
2023-07-12 09:41:01.348 ERROR (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Failed to update device 289C6E20F72C, Error: Invalid token, Flow: Fetch data for device 289C6E20F72C, Line: 227

It retries every 30 seconds, but still has the same error

The sensor data is frozen at the last valid data read, so it still says the API is connected.

When I reload the integration, everything works fine again for the next 1-2 days then it happens again.

Thanks for your help

**Comment by elad-bar** [comment]
Latest is [v3.0.19](https://github.com/radical-squared/aquatemp/releases/tag/v3.0.19) not 3.0.17,
It was fixed in that version.
Going to release another version to fix the same issue when performing an action later today / tomorrow

**Comment by jonml** [comment]
Great, thanks very much

**Comment by davidjac** [comment]
> > > Get to the other account, you will get in app message that will ask you to accept the device invitation
> > 
> > 
> > You mean I should recieve an email to the shared account? Or do I set up an in app account using that new email address. If the former I get no message.
> 
> The entire process:
> 
> * Create account 1
> * Login
> * Logout
> * Create account 2
> * Login
> * Logout
> * Login to account 1
> * Add device
> * Share device to the email of account 2
> * Logout
> * Login to account 2
> * Go to messages
> * Confirm invitation from account 1

Wow that's not obvious or documented. Thanks so much.

**Comment by elad-bar** [comment]
mentioned it (without the steps) in the [readme](https://github.com/radical-squared/aquatemp#shared-device-to-avoid-token-expiration-when-using-the-app-parallel-to-integration), will add also the steps.

**Comment by jonml** [comment]
Hi,
I’m running v3.0.19, but unfortunately it failed again after 18 hour of working fine.

here is the log.  It repeats every 15 minutes.  I forgot to get the debug log set up before I reloaded the integration, but will next time it stops

2023-07-13 14:49:05.302 ERROR (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Failed to fetch parameters, Error: 请重新登录
2023-07-13 14:49:05.330 ERROR (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Failed to update (Attempt #3), Error: 'NoneType' object has no attribute 'get', Line: 143
2023-07-13 14:49:35.461 ERROR (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Failed to fetch parameters, Error: 请重新登录

Thanks for your help

**Comment by dodg3r** [comment]
It doesn't work for me either.
Now I can't log into the app anymore either. It just says "login error"
When I try to change the password it says the same thing.
I think my account has been blocked.

The chinese text in your log means "login again".

**Comment by elad-bar** [comment]
> Hi, I’m running v3.0.19, but unfortunately it failed again after 18 hour of working fine.
> 
> here is the log. It repeats every 15 minutes. I forgot to get the debug log set up before I reloaded the integration, but will next time it stops
> 
> 2023-07-13 14:49:05.302 ERROR (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Failed to fetch parameters, Error: 请重新登录 2023-07-13 14:49:05.330 ERROR (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Failed to update (Attempt #3), Error: 'NoneType' object has no attribute 'get', Line: 143 2023-07-13 14:49:35.461 ERROR (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Failed to fetch parameters, Error: 请重新登录
> 
> Thanks for your help

please try [v3.0.21](https://github.com/radical-squared/aquatemp/releases/tag/v3.0.21), it should address that issue

thanks

**Comment by elad-bar** [comment]
> It doesn't work for me either. Now I can't log into the app anymore either. It just says "login error" When I try to change the password it says the same thing. I think my account has been blocked.
> 
> The chinese text in your log means "login again".

add the same issue today, after several minutes it got fixed, please try again in the app and then in the integration,
please also update to latest v3.0.21

thanks

**Comment by jonml** [comment]
Hi,
I’ve upgraded to v3.0.21, but unfortunately the same issue is still there. It ran for nearly exactly 24 hours, and then stopped receiving updates - although the api flag still shows ‘connected’.

here is a debug log:

2023-07-15 06:42:32.602 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Starting to update device: 289C6E20F72C
2023-07-15 06:42:32.990 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Request to https://cloud.linked-go.com:449/crmservice/api/app/device/getDataByCode, Body: {'deviceCode': '289C6E20F72C', 'protocalCodes': ['Mode', 'Power', 'Mode', 'Manual-mute', 'Set_Temp', '2074', '2075', '2076', '2077', '1158', '1159', '2064', 'ver', 'software_code', 'D01', 'D01-bar', 'D02', 'D03', 'D04', 'D06', 'D07', 'D08', 'D08-bar', 'D09', 'D10', 'D10-bar', 'E01', 'E02', 'E03', 'E04', 'E05', 'E06', 'E07', 'E09', 'E10', 'E11', 'E12', 'F01', 'F02', 'F02-bar', 'F03', 'F03-bar', 'F04', 'F04-bar', 'F05', 'F05-bar', 'F06', 'F06-bar', 'F07', 'F07-bar', 'F10', 'F11', 'F11-r', 'F12', 'F12-r', 'F13', 'F13-r', 'F14', 'F15', 'F16', 'F16-r', 'F17', 'F18', 'F19', 'F19-r', 'F20', 'H01', 'H02', 'H03', 'H06', 'H07', 'H08', 'H09', 'H10', 'H11', 'H12', 'H13', 'H14', 'H15', 'H16', 'H17', 'H18', 'H19', 'H20', 'H21', 'H22', 'H23', 'H24', 'H25', 'H26', 'H27', 'H28', 'H29', 'H30', 'H31', 'H32', 'H33', 'H34', 'H35', 'H36', 'H37', 'H38', 'H39', 'H40', 'H41', 'H42', 'H43', 'O01', 'O02', 'O03', 'O04', 'O05', 'O06', 'O07', 'O08', 'O09', 'P01', 'P02', 'P03', 'P04', 'P05', 'P06', 'P07', 'P08', 'P09', 'R01', 'R02', 'R03', 'R04', 'R05', 'R08', 'R09', 'R10', 'R11', 'R12', 'S01', 'S02', 'S03', 'S04', 'S05', 'S06', 'T01', 'T02', 'T03', 'T04', 'T05', 'T06', 'T07', 'T08', 'T09', 'T10', 'T11', 'T12', 'T13', 'T14', 'T15', 'T16', 'T17', 'T18', 'T19', 'T20', 'T21', 'T22', 'T23', 'T24', 'T25', 'T26', 'T27', 'T28']}, Result: {'error_msg': '请重新登录', 'error_msg_code': '', 'isReusltSuc': False, 'objectResult': None, 'error_code': '-100'}
2023-07-15 06:42:32.990 ERROR (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Failed to fetch parameters, Error: 请重新登录
2023-07-15 06:42:33.020 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Request to https://cloud.linked-go.com:449/crmservice/api/app/device/getDeviceStatus, Body: {'deviceCode': '289C6E20F72C'}, Result: {'error_msg': '请重新登录', 'error_msg_code': '', 'isReusltSuc': False, 'objectResult': None, 'error_code': '-100'}

thanks for your help

**Comment by elad-bar** [comment]
i see only debug messages, do you see error / warning messages with latest version?

thanks

**Comment by jonml** [comment]
Hi,
Here is the full set of logs received for the last connection attempt. Is there anywhere else you would like me to search for error/warning messages?  I don’t see any in ‘notifications’.  The integration ‘fault’ sensor reports ‘ok’ continuously.

thanks


2023-07-15 06:59:13.024 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Starting to update device: 289C6E20F72C
2023-07-15 06:59:13.369 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Request to https://cloud.linked-go.com:449/crmservice/api/app/device/getDataByCode, Body: {'deviceCode': '289C6E20F72C', 'protocalCodes': ['Mode', 'Power', 'Mode', 'Manual-mute', 'Set_Temp', '2074', '2075', '2076', '2077', '1158', '1159', '2064', 'ver', 'software_code', 'D01', 'D01-bar', 'D02', 'D03', 'D04', 'D06', 'D07', 'D08', 'D08-bar', 'D09', 'D10', 'D10-bar', 'E01', 'E02', 'E03', 'E04', 'E05', 'E06', 'E07', 'E09', 'E10', 'E11', 'E12', 'F01', 'F02', 'F02-bar', 'F03', 'F03-bar', 'F04', 'F04-bar', 'F05', 'F05-bar', 'F06', 'F06-bar', 'F07', 'F07-bar', 'F10', 'F11', 'F11-r', 'F12', 'F12-r', 'F13', 'F13-r', 'F14', 'F15', 'F16', 'F16-r', 'F17', 'F18', 'F19', 'F19-r', 'F20', 'H01', 'H02', 'H03', 'H06', 'H07', 'H08', 'H09', 'H10', 'H11', 'H12', 'H13', 'H14', 'H15', 'H16', 'H17', 'H18', 'H19', 'H20', 'H21', 'H22', 'H23', 'H24', 'H25', 'H26', 'H27', 'H28', 'H29', 'H30', 'H31', 'H32', 'H33', 'H34', 'H35', 'H36', 'H37', 'H38', 'H39', 'H40', 'H41', 'H42', 'H43', 'O01', 'O02', 'O03', 'O04', 'O05', 'O06', 'O07', 'O08', 'O09', 'P01', 'P02', 'P03', 'P04', 'P05', 'P06', 'P07', 'P08', 'P09', 'R01', 'R02', 'R03', 'R04', 'R05', 'R08', 'R09', 'R10', 'R11', 'R12', 'S01', 'S02', 'S03', 'S04', 'S05', 'S06', 'T01', 'T02', 'T03', 'T04', 'T05', 'T06', 'T07', 'T08', 'T09', 'T10', 'T11', 'T12', 'T13', 'T14', 'T15', 'T16', 'T17', 'T18', 'T19', 'T20', 'T21', 'T22', 'T23', 'T24', 'T25', 'T26', 'T27', 'T28']}, Result: {'error_msg': '请重新登录', 'error_msg_code': '', 'isReusltSuc': False, 'objectResult': None, 'error_code': '-100'}
2023-07-15 06:59:13.369 ERROR (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Failed to fetch parameters, Error: 请重新登录
2023-07-15 06:59:13.402 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Request to https://cloud.linked-go.com:449/crmservice/api/app/device/getDeviceStatus, Body: {'deviceCode': '289C6E20F72C'}, Result: {'error_msg': '请重新登录', 'error_msg_code': '', 'isReusltSuc': False, 'objectResult': None, 'error_code': '-100'}
2023-07-15 06:59:13.403 ERROR (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Failed to update (Attempt #3), Error: 'NoneType' object has no attribute 'get', Line: 144
2023-07-15 06:59:13.409 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Target temp PC R02, HA Mode: heat
2023-07-15 06:59:13.410 DEBUG (MainThread) [custom_components.aqua_temp.climate] _attr_hvac_mode: heat
2023-07-15 06:59:13.410 DEBUG (MainThread) [custom_components.aqua_temp.climate] _attr_target_temperature: 30.0
2023-07-15 06:59:13.410 DEBUG (MainThread) [custom_components.aqua_temp.climate] _attr_fan_mode: auto
2023-07-15 06:59:13.410 DEBUG (MainThread) [custom_components.aqua_temp.climate] _attr_min_temp: 8.0
2023-07-15 06:59:13.410 DEBUG (MainThread) [custom_components.aqua_temp.climate] _attr_max_temp: 35.0
2023-07-15 06:59:43.482 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Starting to update device: 289C6E20F72C
2023-07-15 06:59:43.575 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Request to https://cloud.linked-go.com:449/crmservice/api/app/device/getDataByCode, Body: {'deviceCode': '289C6E20F72C', 'protocalCodes': ['Mode', 'Power', 'Mode', 'Manual-mute', 'Set_Temp', '2074', '2075', '2076', '2077', '1158', '1159', '2064', 'ver', 'software_code', 'D01', 'D01-bar', 'D02', 'D03', 'D04', 'D06', 'D07', 'D08', 'D08-bar', 'D09', 'D10', 'D10-bar', 'E01', 'E02', 'E03', 'E04', 'E05', 'E06', 'E07', 'E09', 'E10', 'E11', 'E12', 'F01', 'F02', 'F02-bar', 'F03', 'F03-bar', 'F04', 'F04-bar', 'F05', 'F05-bar', 'F06', 'F06-bar', 'F07', 'F07-bar', 'F10', 'F11', 'F11-r', 'F12', 'F12-r', 'F13', 'F13-r', 'F14', 'F15', 'F16', 'F16-r', 'F17', 'F18', 'F19', 'F19-r', 'F20', 'H01', 'H02', 'H03', 'H06', 'H07', 'H08', 'H09', 'H10', 'H11', 'H12', 'H13', 'H14', 'H15', 'H16', 'H17', 'H18', 'H19', 'H20', 'H21', 'H22', 'H23', 'H24', 'H25', 'H26', 'H27', 'H28', 'H29', 'H30', 'H31', 'H32', 'H33', 'H34', 'H35', 'H36', 'H37', 'H38', 'H39', 'H40', 'H41', 'H42', 'H43', 'O01', 'O02', 'O03', 'O04', 'O05', 'O06', 'O07', 'O08', 'O09', 'P01', 'P02', 'P03', 'P04', 'P05', 'P06', 'P07', 'P08', 'P09', 'R01', 'R02', 'R03', 'R04', 'R05', 'R08', 'R09', 'R10', 'R11', 'R12', 'S01', 'S02', 'S03', 'S04', 'S05', 'S06', 'T01', 'T02', 'T03', 'T04', 'T05', 'T06', 'T07', 'T08', 'T09', 'T10', 'T11', 'T12', 'T13', 'T14', 'T15', 'T16', 'T17', 'T18', 'T19', 'T20', 'T21', 'T22', 'T23', 'T24', 'T25', 'T26', 'T27', 'T28']}, Result: {'error_msg': '请重新登录', 'error_msg_code': '', 'isReusltSuc': False, 'objectResult': None, 'error_code': '-100'}
2023-07-15 06:59:43.575 ERROR (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Failed to fetch parameters, Error: 请重新登录
2023-07-15 06:59:43.601 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Request to https://cloud.linked-go.com:449/crmservice/api/app/device/getDeviceStatus, Body: {'deviceCode': '289C6E20F72C'}, Result: {'error_msg': '请重新登录', 'error_msg_code': '', 'isReusltSuc': False, 'objectResult': None, 'error_code': '-100'}

**Comment by elad-bar** [comment]
@jonml sorry it takes time to understand the issue, since i'm not experiencing it, I cannot reproduce it and rely only on your logs,
I think I understand the problem, I tried that the sensor will be in the device level and most probably not extracting it correctly, working on it

**Comment by elad-bar** [comment]
please check [v3.0.22](https://github.com/radical-squared/aquatemp/releases/tag/v3.0.22), hope it will solve the issue

thanks

**Comment by jonml** [comment]
Hi,
Thanks very much for all the work.

I've loaded v3.0.22, and it is running.  Will keep you posted over the next few days

Thanks
________________________________

**Comment by elad-bar** [comment]
great, thanks

**Comment by jonml** [comment]
hi,
sorry, it stopped working again this morning.  almost exactly 24 hours after setting it up - is there some renewal that happens daily?

Here are the debug logs of what i think is the last successful, and 1st unsuccessful calls, if that helps?

thanks for your help

2023-07-16 09:14:24.927 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Request to https://cloud.linked-go.com:449/crmservice/api/app/device/getDeviceStatus, Body: {'deviceCode': '289C6E20F72C'}, Result: {'sessionid': None, 'error_code': '0', 'error_msg': 'Success', 'error_msg_code': '', 'totalSize': None, 'totalPage': None, 'nextPage': None, 'objectResult': {'is_fault': False, 'isFault': False, 'status': 'ONLINE'}, 'isReusltSuc': True}
2023-07-16 09:14:24.927 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_coordinator] Finished fetching Aqua Temp data in 1.452 seconds (success: True)
2023-07-16 09:14:24.929 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Target temp PC R02, HA Mode: heat
2023-07-16 09:14:24.929 DEBUG (MainThread) [custom_components.aqua_temp.climate] _attr_hvac_mode: heat
2023-07-16 09:14:24.929 DEBUG (MainThread) [custom_components.aqua_temp.climate] _attr_target_temperature: 30.0
2023-07-16 09:14:24.929 DEBUG (MainThread) [custom_components.aqua_temp.climate] _attr_fan_mode: auto
2023-07-16 09:14:24.929 DEBUG (MainThread) [custom_components.aqua_temp.climate] _attr_min_temp: 8.0
2023-07-16 09:14:24.930 DEBUG (MainThread) [custom_components.aqua_temp.climate] _attr_max_temp: 35.0
2023-07-16 09:14:54.473 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Starting to update device: 289C6E20F72C
2023-07-16 09:14:54.600 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Request to https://cloud.linked-go.com:449/crmservice/api/app/device/getDataByCode, Body: {'deviceCode': '289C6E20F72C', 'protocalCodes': ['Mode', 'Power', 'Mode', 'Manual-mute', 'Set_Temp', '2074', '2075', '2076', '2077', '1158', '1159', '2064', 'ver', 'software_code', 'D01', 'D01-bar', 'D02', 'D03', 'D04', 'D06', 'D07', 'D08', 'D08-bar', 'D09', 'D10', 'D10-bar', 'E01', 'E02', 'E03', 'E04', 'E05', 'E06', 'E07', 'E09', 'E10', 'E11', 'E12', 'F01', 'F02', 'F02-bar', 'F03', 'F03-bar', 'F04', 'F04-bar', 'F05', 'F05-bar', 'F06', 'F06-bar', 'F07', 'F07-bar', 'F10', 'F11', 'F11-r', 'F12', 'F12-r', 'F13', 'F13-r', 'F14', 'F15', 'F16', 'F16-r', 'F17', 'F18', 'F19', 'F19-r', 'F20', 'H01', 'H02', 'H03', 'H06', 'H07', 'H08', 'H09', 'H10', 'H11', 'H12', 'H13', 'H14', 'H15', 'H16', 'H17', 'H18', 'H19', 'H20', 'H21', 'H22', 'H23', 'H24', 'H25', 'H26', 'H27', 'H28', 'H29', 'H30', 'H31', 'H32', 'H33', 'H34', 'H35', 'H36', 'H37', 'H38', 'H39', 'H40', 'H41', 'H42', 'H43', 'O01', 'O02', 'O03', 'O04', 'O05', 'O06', 'O07', 'O08', 'O09', 'P01', 'P02', 'P03', 'P04', 'P05', 'P06', 'P07', 'P08', 'P09', 'R01', 'R02', 'R03', 'R04', 'R05', 'R08', 'R09', 'R10', 'R11', 'R12', 'S01', 'S02', 'S03', 'S04', 'S05', 'S06', 'T01', 'T02', 'T03', 'T04', 'T05', 'T06', 'T07', 'T08', 'T09', 'T10', 'T11', 'T12', 'T13', 'T14', 'T15', 'T16', 'T17', 'T18', 'T19', 'T20', 'T21', 'T22', 'T23', 'T24', 'T25', 'T26', 'T27', 'T28']}, Result: {'error_msg': '请重新登录', 'error_msg_code': '', 'isReusltSuc': False, 'objectResult': None, 'error_code': '-100'}
2023-07-16 09:14:54.600 ERROR (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Failed to fetch parameters, Error: 请重新登录
2023-07-16 09:14:54.633 DEBUG (MainThread) [custom_components.aqua_temp.managers.aqua_temp_api] Request to https://cloud.linked-go.com:449/crmservice/api/app/device/getDeviceStatus, Body: {'deviceCode': '289C6E20F72C'}, Result: {'error_msg': '请重新登录', 'error_msg_code': '', 'isReusltSuc': False, 'objectResult': None, 'error_code': '-100'}

**Comment by jonml** [comment]
Hi, it stopped again after exactly 24 hours from its reload.  
I have updated to v3.0.23, and reloaded it again

thanks
Jon

**Comment by elad-bar** [comment]
I think I found it, there are configuration files for each api type, for some reason, I have missed to set up the configuration of how to extract the error_code, that's the reason it didn't match the criteria of error code == -100 and didn't renew the token, I have fixed it in [v3.0.24](https://github.com/radical-squared/aquatemp/releases/tag/v3.0.24), pls check and let me know whether it solved the issue

**Comment by jonml** [comment]
Great, thanks.
I’ve installed v3.0.24
🤞
Will let you know what happens in 24 hours!
Cheers
Jon

**Comment by elad-bar** [comment]
@jonml any updates?

Thanks

**Comment by jonml** [comment]
Hi, yes it is all good. Looks like it is fixed.
Thanks very much!
________________________________

**Comment by elad-bar** [comment]
Great!
closing the issue, thanks
